# Expoloratory Data Analysis (EDA)

Разведочный анализ данных. Нахождение закономерностей и проблем в исходном датасете.


## Загрузка данных

Загрузим исходные данные.

In [49]:
import pandas as pd

path = "../data/raw/android_games_eda_ready.csv"
df = pd.read_csv(path)
df.head()

,game_id,game_name,developer_name,package_name,genre,sub_genre,monetization_model,contains_ads,has_in_app_purchases,multiplayer_support,...,has_seasonal_event,cross_platform_available,cloud_save_support,is_hit_game,post_30d_revenue_usd,post_30d_rating_count,featured_duration_days,store_platform,country_code_primary,row_checksum_id
0,622,Game_622_Arena,Studio_205,com.studio1664.game622,Casual,Party Game,Freemium,Yes,Yes,NaN,...,False,True,True,0,9963.75,6328,NaN,Android,IN,CHK_100621
1,4386,Game_4386_Rush,Studio_923,com.studio1615.game4386,RPG,Idle RPG,Free,Yes,No,Online,...,False,True,False,0,17880.43,9346,NaN,Android,IN,CHK_104385
2,6795,Game_6795_Tycoon,Studio_921,com.studio44.game6795,RPG,Idle RPG,Free,Yes,No,NaN,...,True,False,False,0,6060.03,4096,7.0,Android,BR,CHK_106794
3,4563,Game_4563_Quest,Studio_1873,com.studio1353.game4563,Casual,Idle Clicker,Paid,No,Yes,NaN,...,False,False,False,0,816621.72,12402,NaN,Android,US,CHK_104562
4,9265,Game_9265_Legends,Studio_1608,com.studio1308.game9265,RPG,MMORPG,Free,Yes,No,NaN,...,True,False,True,0,159839.20,7517,NaN,Android,IN,CHK_109264


## Первичный анализ данных

Проанализируем данные и выявим:
* общую информацию о размере
* типы данных и количество уникальных
* основные статистики числовых признаков
* пропуски и дубликаты

In [50]:
import numpy as np

### Размер датасета

In [51]:
print(f"Строк: {df.shape[0]}, столбцов: {df.shape[1]}")

Строк: 10150, столбцов: 57


### Типы данных

In [52]:
print(df.dtypes.value_counts())

str        23
float64    21
int64       8
bool        5
Name: count, dtype: int64


### Количество уникальных значений для каждого столбца

In [53]:
for col in df.columns:
    uniq = df[col].nunique()
    print(f"{col}: {uniq} уникальных")

game_id: 10000 уникальных
game_name: 10000 уникальных
developer_name: 2222 уникальных
package_name: 10000 уникальных
genre: 8 уникальных
sub_genre: 31 уникальных
monetization_model: 4 уникальных
contains_ads: 2 уникальных
has_in_app_purchases: 2 уникальных
multiplayer_support: 3 уникальных
age_rating: 3 уникальных
art_style: 5 уникальных
engine_used: 4 уникальных
region_focus: 6 уникальных
publisher_tier: 4 уникальных
event_theme: 5 уникальных
release_date: 2819 уникальных
soft_launch_date: 2872 уникальных
last_update_date: 3140 уникальных
was_featured_on_store: 2 уникальных
featured_start_date: 2029 уникальных
featured_end_date: 1988 уникальных
price_usd: 688 уникальных
downloads: 9855 уникальных
retention_day1_pct: 3463 уникальных
retention_day7_pct: 2581 уникальных
retention_day30_pct: 1673 уникальных
active_users_30d: 8250 уникальных
avg_session_minutes: 2972 уникальных
avg_daily_sessions: 575 уникальных
crash_rate_pct: 661 уникальных
apk_size_mb: 8310 уникальных
update_frequency_d

### Числовые признаки

In [54]:
display(df.describe(include=[np.number]).T)

,count,mean,std,min,25%,50%,75%,max
game_id,10150.0,4999.301970,2.886833e+03,1.00,2502.2500,4997.500,7501.7500,10000.00
price_usd,10150.0,0.699186,1.805539e+00,-4.94,0.0000,0.000,0.0000,24.70
downloads,10150.0,319704.915862,1.265406e+06,1061.00,46301.0000,110428.500,267512.2500,53450424.00
retention_day1_pct,10150.0,30.215609,9.157124e+00,5.00,23.7000,29.550,36.1700,68.74
retention_day7_pct,10150.0,16.693777,6.229105e+00,2.00,12.1200,15.900,20.5200,47.48
retention_day30_pct,10150.0,7.945300,3.761827e+00,1.00,5.1700,7.280,10.0000,25.00
active_users_30d,10150.0,20166.194778,4.418765e+04,62.00,2894.5000,7482.500,19502.7500,1020550.00
avg_session_minutes,10150.0,19.627916,7.686156e+00,3.00,14.2800,19.570,24.8900,47.48
avg_daily_sessions,10150.0,3.342308,1.021145e+00,0.50,2.6600,3.350,4.0200,7.17
crash_rate_pct,10150.0,2.525782,1.430158e+00,-2.78,1.4900,2.495,3.5000,8.83


### Пропуски

In [55]:
missing = df.isnull().sum()
missing_pct = (missing / len(df)) * 100
missing_df = pd.DataFrame({'Пропусков': missing, '%': missing_pct})
missing_df = missing_df[missing_df['Пропусков'] > 0].sort_values('Пропусков', ascending=False)
display(missing_df)

,Пропусков,%
event_theme,7682,75.684729
featured_duration_days,6984,68.807882
featured_end_date,6808,67.073892
featured_start_date,6805,67.044335
multiplayer_support,4066,40.059113
soft_launch_date,1012,9.970443
marketing_spend_usd,807,7.950739
arppu_usd,708,6.975369
cpi_usd,509,5.014778


### Дубликаты

In [56]:
full_dups = df.duplicated().sum()

print(f"Полных дубликатов строк: {full_dups}")

Полных дубликатов строк: 150


### Выводы
- **Размер**: 10150 строк, 57 столбцов.
- **Типы данных**:
  * 6 булевых (bool)
  * 8 целочисленных (int64)
  * 11 вещественных (float64)
  * 32 — строковые (object)
    - Многие строковые колонки по сути категориальные
- **Пропуски**: 9 столбцов имеют пропуски. Наибольшая доля у `event_theme` (75.7%) и `featured_duration_days` (68.8%).
- **Дубликаты**: 150 полных дубликатов строк - их необходимо удалить.
- **Константные / бесполезные колонки**: `store_platform` (только "Android"), `game_id` и `row_checksum_id` (идентификаторы), `package_name` (10 000 уникальных). Их удаление сократит размерность без потери информации.

---

## Первичная обработка данных

Теперь обработаем данные.

### Удаление дубликатов

In [57]:
df = df.drop_duplicates()
print("Размер после удаления полных дубликатов:", df.shape)

Размер после удаления полных дубликатов: (10000, 57)


### Удаление заведомо неинформативных столбцов

In [58]:
cols_to_drop = ['store_platform', 'row_checksum_id', 'game_id', 'game_name', 'package_name']
df = df.drop(columns=cols_to_drop)
print("Размер после удаления неинформативных столбцов:", df.shape)

Размер после удаления неинформативных столбцов: (10000, 52)


### Исправление типов

#### Bool

Удостоверимся, что колонки имеют тип bool.

In [59]:
bool_cols = ['contains_ads', 'has_in_app_purchases', 'was_featured_on_store',
             'is_weekend_release', 'has_seasonal_event', 'cross_platform_available', 'cloud_save_support']
for col in bool_cols:
    if col in df.columns:
        df[col] = df[col].astype(bool)

#### Category

Категориальные колонки преобразуем в category.

In [60]:
cat_cols = ['genre', 'sub_genre', 'monetization_model', 'multiplayer_support',
            'age_rating', 'art_style', 'engine_used', 'region_focus', 'publisher_tier',
            'event_theme', 'country_code_primary']
existing_cats = [col for col in cat_cols if col in df.columns]
df[existing_cats] = df[existing_cats].astype('category')

#### Datetime

Даты преобразуем в datetime.

In [61]:
date_cols = ['release_date', 'soft_launch_date', 'last_update_date',
             'featured_start_date', 'featured_end_date']
for col in date_cols:
    if col in df.columns:
        df[col] = pd.to_datetime(df[col], errors='coerce')

#### Итог

Теперь данные имеют следующий вид:

In [62]:
print("\nТипы данных после преобразований:")
print(df.dtypes)


Типы данных после преобразований:
developer_name                         str
genre                             category
sub_genre                         category
monetization_model                category
contains_ads                          bool
has_in_app_purchases                  bool
multiplayer_support               category
age_rating                        category
art_style                         category
engine_used                       category
region_focus                      category
publisher_tier                    category
event_theme                       category
release_date                datetime64[us]
soft_launch_date            datetime64[us]
last_update_date            datetime64[us]
was_featured_on_store                 bool
featured_start_date         datetime64[us]
featured_end_date           datetime64[us]
price_usd                          float64
downloads                            int64
retention_day1_pct                 float64
retention_day7_pct 

In [63]:
print("\nПропусков после преобразований дат:")
print(df[date_cols].isnull().sum())


Пропусков после преобразований дат:
release_date              0
soft_launch_date       1000
last_update_date          0
featured_start_date    6697
featured_end_date      6697
dtype: int64


### Удаление утечек

Удаляем признаки, которые содержат данные из будущего (например, `post_30d_revenue_usd` - доход через 30 дней после релиза).

Также удалим столбец `is_hit_game`, который показывает, стала ли игра хитом. Это также не относится к задаче.

In [64]:
leakage_cols = ['post_30d_revenue_usd', 'post_30d_rating_count', 'is_hit_game']

df = df.drop(columns=leakage_cols)
print("Размер после удаления колонок:", df.shape)

Размер после удаления колонок: (10000, 49)


Посмотрим на текущие пропуски:

In [67]:
missing = df.isnull().sum()
missing_pct = (missing / len(df)) * 100
missing_df = pd.DataFrame({'Пропусков': missing, '%': missing_pct})
missing_df = missing_df[missing_df['Пропусков'] > 0].sort_values('Пропусков', ascending=False)
display(missing_df)

,Пропусков,%
event_theme,7577,75.77
featured_duration_days,6871,68.71
featured_end_date,6697,66.97
featured_start_date,6697,66.97
multiplayer_support,4013,40.13
soft_launch_date,1000,10.00
marketing_spend_usd,800,8.00
arppu_usd,700,7.00
cpi_usd,500,5.00


### Обработка пропусков

Обратим внимание, что у нас всё ещё осталось много пропусков:
1. `event_theme` (75%) - почти пустой, не несёт полезной нагрузки. Для простоты лучше удалить.
2. `featured_*` (67%) - пропуски означают, что игру не продвигали в магазине (не "featured").
Вместо заполнения дат создадим один бинарный признак is_featured, равный `1`, если даты не пусты, иначе `0`.
3. `multiplayer_support` (40%) — категориальный признак с тремя значениями ("Online", "Offline", и NaN). Добавим новую категорию "Unknown" вместо пропусков. Это сохранит информацию о том, что режим мультиплеера не указан.
4. `soft_launch_date` (10%) — для игр без раннего релиза даты нет. Создадим бинарный флаг `had_soft_launch` и числовой признак days_soft_to_release (разница между релизом и ранним релизом). Изначальную колонку удалим.
5. Финансовые `marketing_spend_usd`, `cpi_usd`, `arppu_usd` (5–8%) — пропуски умеренные. Заполним медианой, так как распределения скошены.

В конце проверим, что пропусков не осталось.

1. Удаляем столбец `event_time`.

In [68]:
df = df.drop(columns=['event_theme'])

2. Признак is_featured (было ли продвижение) и заполнение featured_duration_days